extensions/amp-date-display/0.2/date-display.js

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {useResourcesNotify} from '../../../src/preact/utils';
+import {userAssert} from '../../../src/log';
+
+/** @const {string} */
+const DEFAULT_LOCALE = 'en';
+
+/** @const {number} */
+const DEFAULT_OFFSET_SECONDS = 0;
+
+/** @typedef {{
+  year: number,
+  month: number,
+  monthName: string,
+  monthNameShort: string,
+  day: number,
+  dayName: string,
+  dayNameShort: string,
+  hour: number,
+  minute: number,
+  second: number,
+  iso: string,
+}} */
+let VariablesV2Def;
+
+/** @typedef {{
+  year: number,
+  month: number,
+  monthName: string,
+  monthNameShort: string,
+  day: number,
+  dayName: string,
+  dayNameShort: string,
+  hour: number,
+  minute: number,
+  second: number,
+  iso: string,
+  yearTwoDigit: string,
+  monthTwoDigit: string,
+  dayTwoDigit: string,
+  hourTwoDigit: string,
+  hour12: string,
+  hour12TwoDigit: string,
+  minuteTwoDigit: string,
+  secondTwoDigit: string,
+  dayPeriod: string,
+ }} */
+let EnhancedVariablesV2Def;
+
+/**
+ * @param {!JsonObject} props
+ * @return {Preact.Renderable}
+ */
+export function DateDisplay(props) {
+  const render = props['render'];
```

Comment:
Continuing https://github.com/ampproject/amphtml/pull/25969#discussion_r360539255:

> But props are really meant to be compilable symbols.

We cannot change this. If we allow non-string access, then mangling will break the components between the amp-to-react boundary. Eg, we retrieve the `display-in` attribute, and set it to `props['displayIn']` as part of collect props. But Closure will mangle the `const { displayIn } = props` that it receives in the Preact component (to `QD`, in my case)`. So we get `undefined` instead of the value we needed.

```js
// Collect props does this:
const props = { ['displayIn']: el.getAttribute('display-in')};

// Preact does this due to mangling:
const displayIn = props.QD;
```

>  Even further, I've recently seen mainly the following form used and find it very helpful:

We can still do destructuring, if you'd like:

```js
const { 'render': render } = props
```

It just has to use string keys and not the shorthand.

Reply:
Yeah, I see your point. But it's so unidiomatic. What's our long-term approach here? Can we instruct closure to ignore minification for props? For these modules in their entirety? Should we transform these modules to `{render}` to `{'render': render}` before feeding them to Closure? I.e. let's take the step we need now, but what's our north star here?

Reply:
I can see two solutions:

1. We transform any `props` access to string form before closure. This has the downside that any `foo(props)` call must also use `props` as the parameter name, which is extremely difficult to lint for. We may be able to get away with it if we force the type to still be a `!JsonObject`, but then our devs will sometimes have to write string keys for `JsonObject`s and sometimes not.

2. We extern the props definition for every component. I think this is actually the better solution. It just requires us to locate the extern file somewhere else, which is annoying.

Reply:
IMHO, I agree. I think either solution (even maybe converting d.ts to externs) are good for what we want. React-style is a bit of a different domain language, especially if we allow jsx. So special transformers are not a crazy idea for certain.